src/main/java/org/venice/piazza/idam/util/GxOAuthClient.java

```diff
@@ -0,0 +1,159 @@
+		String accessToken = tokenResponse.getBody();
+		JacksonJsonParser parser = new JacksonJsonParser();
+		Map<String, Object> accessTokenMap = parser.parseMap(accessToken);
+		return accessTokenMap.get("access_token").toString();
+	}
+
+	public ResponseEntity<GxOAuthResponse> getGxUserProfile(final String accessToken) throws HttpClientErrorException, HttpServerErrorException {
+		HttpHeaders headers = new HttpHeaders();
+		headers.set(AUTHORIZATION, "Bearer " + accessToken);
+		ResponseEntity<GxOAuthResponse> profileResponse = new ResponseEntity<>(restTemplate.exchange(
+				getProfileRequestUrl(),
+				HttpMethod.GET,
+				new HttpEntity<>("parameters", headers),
+				GxOAuthResponse.class).getBody(),
+				HttpStatus.OK);
+		return profileResponse;
+	}
+
+	public UserProfile getUserProfileFromGxProfile(GxOAuthResponse oAuthResponse) throws InvalidNameException {
+		final LdapName ldapName = new LdapName(oAuthResponse.getDn());
+		final Rdn countryRdn = ldapName.getRdns().stream().filter(rdn ->
+				rdn.getType().equalsIgnoreCase("C")
+		).
+				findAny().
+				orElse(null);
+		String country = "";
+		if (countryRdn != null) {
+			country = countryRdn.getValue().toString();
+		}
+
+		String dutyCode;
+		String adminCode;
+		if ("NGA".equalsIgnoreCase(oAuthResponse.getServiceOrAgency())) {
+			dutyCode = oAuthResponse.getServiceOrAgency();
+			adminCode = oAuthResponse.getServiceOrAgency();
+		} else {
+			dutyCode = oAuthResponse.getAdministrativeOrganizationCode();
+			adminCode = oAuthResponse.getAdministrativeOrganizationCode();
+		}
+		final UserProfile profile = new UserProfile();
+		profile.setDistinguishedName(oAuthResponse.getDn());
+		profile.setUsername(oAuthResponse.getUsername());
+		profile.setCountry(country);
+		profile.setDutyCode(dutyCode);
+		profile.setAdminCode(adminCode);
+		return profile;
+	}
+
+	public String getOAuthUrlForGx(final String redirectUri)  {
+		return "http://" + gxBaseUrl + "/ms_oauth/oauth2/endpoints/oauthservice/authorize?scope=UserProfile.me&client_id="
+				+ gxClientId + "&response_type=code&redirect_uri=" + redirectUri;
+	}
+
+	public String getLogoutUrl()  {
+		return "http://" + gxBaseUrl + "/oam/server/logout";
+	}
+
+	public String getRedirectUri(HttpServletRequest request) {
+		return getRequestBaseUrl(request) + "/login";
+	}
+
+	public String getUiUrl(HttpServletRequest request) {
+		return getRequestBaseUrl(request).replace("pz-idam", "beachfront");
+	}
+
+	private String getRequestBaseUrl(HttpServletRequest request) {
+		return request.getScheme() + "://" +
+				request.getServerName() + ":" + request.getServerPort();
+	}
+
+	private String getTokenRequestUrl(
+			final String redirectUri,
+			final String code) {
+		return "http://" + gxBaseUrl + "/ms_oauth/oauth2/endpoints/oauthservice/tokens?grant_type=authorization_code&redirect_uri="
+				+ redirectUri + "&code=" + code;
+	}
+
+	private String getProfileRequestUrl() {
+		return "http://" + gxBaseUrl + "/ms_oauth/resources/userprofile/me";
```

Comment:
We should make this a property / variable.

Reply:
So we should just move this to the top as a private data member?

Reply:
I was thinking a property file or environment variable, since these contexts are determined by an external entity (GEOAxIS) and could change. We don't want to have to recompile every time they decide to change their URLs.